src/datatype.rs

```diff
@@ -749,201 +752,140 @@ fn lookup_global_promote(lookup_name: &str) -> Option<(Args, Datatype)> {
-            LookupResult::WrongType
-        }
+    if ltype == Datatype::Unknown {
+        LookupResult::Found(possible_args.unwrap(), possible_rtype.unwrap())
     } else {
-        LookupResult::NotFound
+        // If it was the right type, it would already have been returned as `Found`, above.
+        LookupResult::WrongType
     }
 }
 
-fn lookup_promote(lookup_name: &str, ltype: Datatype) -> LookupResult {
-    let promotes = match Game::game() {
+fn promote(lookup_name: &str) -> Option<&Vec<(Datatype, Args, Datatype)>> {
+    let promotes_map = match Game::game() {
         #[cfg(feature = "ck3")]
-        Game::Ck3 => crate::ck3::tables::datafunctions::PROMOTES,
+        Game::Ck3 => &crate::ck3::tables::datafunctions::PROMOTES_MAP,
         #[cfg(feature = "vic3")]
-        Game::Vic3 => crate::vic3::tables::datafunctions::PROMOTES,
+        Game::Vic3 => &crate::vic3::tables::datafunctions::PROMOTES_MAP,
         #[cfg(feature = "imperator")]
-        Game::Imperator => crate::imperator::tables::datafunctions::PROMOTES,
+        Game::Imperator => &crate::imperator::tables::datafunctions::PROMOTES_MAP,
     };
-    lookup_promote_or_function(lookup_name, ltype, promotes)
+    promotes_map.get(lookup_name)
 }
 
-fn lookup_function(lookup_name: &str, ltype: Datatype) -> LookupResult {
-    let functions = match Game::game() {
+fn lookup_promote(lookup_name: &str, ltype: Datatype) -> LookupResult {
+    let result = promote(lookup_name);
+    result.map_or(LookupResult::NotFound, |x| lookup_promote_or_function(ltype, x))
+}
+
+fn function(lookup_name: &str) -> Option<&Vec<(Datatype, Args, Datatype)>> {
+    let functions_map = match Game::game() {
         #[cfg(feature = "ck3")]
-        Game::Ck3 => crate::ck3::tables::datafunctions::FUNCTIONS,
+        Game::Ck3 => &crate::ck3::tables::datafunctions::FUNCTIONS_MAP,
         #[cfg(feature = "vic3")]
-        Game::Vic3 => crate::vic3::tables::datafunctions::FUNCTIONS,
+        Game::Vic3 => &crate::vic3::tables::datafunctions::FUNCTIONS_MAP,
         #[cfg(feature = "imperator")]
-        Game::Imperator => crate::imperator::tables::datafunctions::FUNCTIONS,
+        Game::Imperator => &crate::imperator::tables::datafunctions::FUNCTIONS_MAP,
     };
-    lookup_promote_or_function(lookup_name, ltype, functions)
+    functions_map.get(lookup_name)
+}
+
+fn lookup_function(lookup_name: &str, ltype: Datatype) -> LookupResult {
+    let result = function(lookup_name);
+    result.map_or(LookupResult::NotFound, |x| lookup_promote_or_function(ltype, x))
+}
+
+pub struct CaseInsensitiveStr(pub(crate) &'static str);
+
+impl PartialEq for CaseInsensitiveStr {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.eq_ignore_ascii_case(other.0)
+    }
+}
+
+impl Eq for CaseInsensitiveStr {}
+
+impl std::hash::Hash for CaseInsensitiveStr {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.0.to_ascii_lowercase().hash(state);
```

Comment:
This might be an expensive hash since it allocates a String. How often is it called?

Reply:
This is called for every insertion and every query. I don't think there is anyway to avoid this unfortunately.